pkg/intermediate/aggregate.go

```diff
@@ -0,0 +1,218 @@
+package intermediate
+
+import (
+	"encoding/binary"
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/vmware/go-ipfix/pkg/entities"
+	"github.com/vmware/go-ipfix/pkg/registry"
+)
+
+type aggregation struct {
+	// tupleRecordMap maps each connection with its records
+	tupleRecordMap map[Tuple][]entities.Record
+	// tupleRecordLock allows multiple readers or one writer at the same time
+	tupleRecordLock sync.RWMutex
+	// workerPool is for exchanging worker channels to process message
+	workerPool chan chan *entities.Message
+	// messageChan is the channel to receive message
+	messageChan chan *entities.Message
+	// workerNum is the number of workers to process message
+	workerNum int
+	// workerList is the list of workers
+	workerList []*worker
+}
+
+type Tuple struct {
+	// SourceAddress is the encoded format of sourceIPv4Address
+	SourceAddress uint32
+	// DestinationAddress is the encoded format of destinationIPv4Address
+	DestinationAddress uint32
+	Protocol           uint8
+	SourcePort         uint16
+	DestinationPort    uint16
+}
+
+func InitAggregationProcess(messageChan chan *entities.Message, workerNum int) (*aggregation, error) {
+	if messageChan == nil {
+		return nil, fmt.Errorf("Cannot create aggregation process without message channel.")
+	} else if workerNum <= 0 {
+		return nil, fmt.Errorf("Worker number cannot be <= 0.")
+	}
+	return &aggregation{
+		make(map[Tuple][]entities.Record),
+		sync.RWMutex{},
+		make(chan chan *entities.Message),
+		messageChan,
+		workerNum,
+		make([]*worker, 0),
+	}, nil
+}
+
+func (a *aggregation) Start() {
+	for i := 0; i < a.workerNum; i++ {
+		w := createWorker(i, a.workerPool, a.AggregateMsgBy5Tuple)
+		w.start()
+		a.workerList = append(a.workerList, w)
+	}
+	for message := range a.messageChan {
+		channel := <-a.workerPool
+		channel <- message
+	}
+}
+
+func (a *aggregation) Stop() {
+	for _, worker := range a.workerList {
+		worker.stop()
+	}
+}
+
+// AggregateMsgBy5Tuple gets 5-tuple info from records in message and stores in cache
+func (a *aggregation) AggregateMsgBy5Tuple(message *entities.Message) error {
+	addOriginalExporterInfo(message)
+	if message.Set.GetSetType() == entities.Template { // skip template records
+		return nil
+	}
+	records := message.Set.GetRecords()
+	a.tupleRecordLock.Lock()
+	defer a.tupleRecordLock.Unlock()
+	for _, record := range records {
+		tuple, err := getTupleFromRecord(record)
+		if err != nil {
+			return err
+		}
+		a.correlateRecords(tuple, record)
```

Comment:
What happens if we get multiple records for the same flow from the same source or destination node?